patches/sitecustomize.py

```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 
-# Monkey patches BigQuery client creation to use proxy.
-kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
+    # Use Data Proxy if user has specified to use the Kaggle project, or if
+    # there are no connected GCP accounts (to maintain backwards compatibility).
+    if bq_user_jwt is None and specified_project and specified_project != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = specified_project == 'kaggle' or bq_user_jwt is None
+    if use_data_proxy:
+        if data_proxy_project is None or kaggle_proxy_token is None:
+            # We don't have the data proxy info so leave the bq client unmodified.
+            return bq_client(*args, **kwargs)
+        print("Using Kaggle's public dataset BigQuery integration.")
```

Comment:
This output would be new.  Are we sure it is desirable?  I suppose this goes in the cell output?  If so, is it only done when a bq dataset is attached to the kernel?

Reply:
I mentioned to others that I was thinking of adding this messaging  (https://kaggle.slack.com/archives/CDB736R0C/p1545422371002800?thread_ts=1545422009.001500&cid=CDB736R0C) and they seemed to like  the idea. I don't think a BQ dataset being attached to the Kernel is related here, but correct me if I'm wrong - it should only display when the user instantiates a BigQuery client (ie. client = bigquery.Client(project='KAGGLE'))

Reply:
Ah yes, that seems fine then indeed.  Though IMO the intent should be expressed by the user in the first place, not some magic behind the scenes and then telling the user what the magic did.

Reply:
I agree I'd like the least amount of magic as possible. However we already have some magic going on (Kaggle data proxy injecting our own credentials) and I figured now that we're adding another - less magical way of using it - the user should know which one is being used. My hope is that the user explicitly "tells" us which way they want to use (ie. project = 'mypersonalproject' or 'kaggle') but this only makes it more clear to them what is happening.

Reply:
There is an opportunity to avoid the magic (and hopefully evt get to a place without any magic) here, let's take it.  Project 'kaggle' isn't a real GCP product (or if it is wouldn't be used here).